incident/utils/forms.py

```diff
@@ -0,0 +1,98 @@
+from django import forms
+
+from incident.models import choices
+
+
+class FilterForm(forms.Form):
+
+    def __init__(self, *args, **kwargs):
+
+        data = kwargs.pop('data', None)
+        filters = data.get('filters', None)
+        self.id = data.get('id', '')
+        self.title = data.get('title', '')
+        self.url = data.get('url', '')
+        super().__init__(*args, **kwargs)
+
+        if filters:
+            for item in filters:
+                _type = item.get('type', None)
+                name = item.get('name', '')
+                label = item.get('title', '')
+                field = None
+                kwargs = {
+                    'required': False,
+                    'label': label,
+                }
+
+                if _type == 'text' or _type == 'autocomplete':
+                    field = forms.CharField
+
+                if _type == 'date':
+                    field = forms.DateField
+                    kwargs['widget'] = forms.DateInput
+
+                if _type == 'bool':
+                    field = forms.ChoiceField
+                    kwargs['widget'] = forms.RadioSelect
+                    kwargs['choices'] = [('Yes', 'Yes',), ('No', 'No',)]
+
+                if _type == 'checkbox':
+                    field = forms.MultipleChoiceField
+                    kwargs['widget'] = forms.CheckboxSelectMultiple
+                    kwargs['choices'] = [[x[0], x[1].capitalize()] for x in item.get('choices', [])]
+
+                if _type == 'choice':
+                    field = forms.ChoiceField
+                    kwargs['widget'] = forms.Select
+                    kwargs['choices'] = [[x[0], x[1].capitalize()] for x in item.get('choices', [])]
+
+                if _type == 'radio':
+                    field = forms.ChoiceField
+                    kwargs['widget'] = forms.RadioSelect
+                    kwargs['choices'] = [[x[0], x[1].capitalize()] for x in choices.MAYBE_BOOLEAN]
+
+                if field:
+                    self.fields[name] = field(**kwargs)
+                    field.field_type = _type
+
+
+def get_filter_forms(request, serialized_filters):
+    filter_forms = []
+    categories = []
+
+    for item in serialized_filters:
+
+        # if the item has a filters object then create a form from the item
+        if item.get('filters', []):
```

Comment:
As you mention in your comments on #1218, there is a need a "limit to a category" option, and your instinct is correct that these filters are not explicitly defined in the `filters` key, but what we have here is not quite the logic for determining that (and it's definitely not surprising you didn't infer this is how it works, there's not really documentation on the format of what's returned from `get_serialized_filters`, so no worries there). 

Here's what's happening: every item in `serialized_filters` "belongs to" a category _except_ the "general" filters item. The way to tell the difference is that each item has the `id` key set to the the primary key of the corresponding category page (which you are using to set the value of the appropriate checkbox below), and the general filters have `id` set to `-1`. The items with without filters (or with `filters == []`) are categories without specific filters that belong to them. 

So the code you've written here is very close to the correct behavior, but it will only create category checkboxes for certain categories, namely ones in the latter group without filters. There may be a more clever way to do this, but essentially the `categories` variable is every item in `serialized_filters` with an `id` key that's not `-1`, or, like, 

```python
[item for item in serialized_filters if item.get('id', -1) != -1]
```

Though how you want to express this I think is up to you. Let me know if that makes sense! The code that generates this is here: 

https://github.com/freedomofpress/pressfreedomtracker.us/blob/42bff62725af8b77b8cd220d5c5894b7c168541e/incident/utils/incident_filter.py#L529-L553

Which, though it's not the clearest, you can see the first part is a block of "general filters" and the second block loops through CategoryPages renders the filters that "belong" to each of those.

Reply:
Makes sense now, and the way it's now implemented looks sensible 👍